libopensc, pkcs15init: silence some gcc errors spotted on Ubuntu 23.10

LocutusOfBorg opened the pull request:
development release with gcc-13, lto and O3 optimization level E.g. of error:
libtool: link: gcc -g -O0 -Wall -Wextra -Wno-unused-parameter -Werror -Wstrict-aliasing=2 -g -O2 -ffile-prefix-map=/<<PKGBUILDDIR>>=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -fdebug-prefix-map=/<<PKGBUILDDIR>>=/usr/src/opensc-0.23.0-1 -Wno-error=deprecated-declarations -Wno-error=stringop-overflow -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -Wl,-z -Wl,relro -o fuzz_pkcs15init fuzz_pkcs15init.o fuzzer_reader.o fuzzer.o  ../../../src/libopensc/.libs/libopensc.a -lz -lgio-2.0 -lgobject-2.0 -leac -lcrypto ../../../src/common/.libs/libscdl.a -ldl ../../../src/pkcs15init/.libs/libpkcs15init.a ../../../src/common/.libs/libcompat.a -pthread ../../../src/pkcs15init/pkcs15-lib.c: In function 'sc_pkcs15init_update_any_df': ../../../src/pkcs15init/pkcs15-lib.c:3247:21: error: 'bufsize' may be used uninitialized [-Werror=maybe-uninitialized]
 3247 |                 r = sc_pkcs15init_update_file(profile, p15card, file, buf, bufsize);
      |                     ^
../../../src/pkcs15init/pkcs15-lib.c:3234:25: note: 'bufsize' was declared here
 3234 |         size_t          bufsize;
      |                         ^
lto1: all warnings being treated as errors

<!--
Thank you for your pull request.

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX'
(without quotes) in the commit message.

Mention which card(s) are used during testing. To get the name of your card,
run this command: `opensc-tool -n`
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] Documentation is added or updated
- [ ] New files have a LGPL 2.1 license statement
- [ ] PKCS#11 module is tested
- [ ] Windows minidriver is tested
- [ ] macOS tokend is tested

Jakuje commented on src/pkcs15init/profile.c:
```diff
@@ -499,6 +499,8 @@ sc_profile_get_pin_info(struct sc_profile *profile,
 		int id, struct sc_pkcs15_auth_info *info)
 {
 	struct pin_info	*pi;
+	struct sc_pkcs15_auth_info info_tmp = {0};
+	*info = info_tmp;
```
I do not think this is correct. In case of the failure, you will pass out of the function the pointer to the local variable, which will no longer exist after you `return`. I think you should just initialize the pointer to NULL.

LocutusOfBorg replied:
well, this tmp variable is a no-op, just to please gcc compiler.
When I tried with memset, for some reasons it didn't work, and it works now.
pushed.

Jakuje replied:
Do you have the specific error message you were getting?

In any case, I think we should first check the pointer is not null before dereferencing it.

LocutusOfBorg replied:
sorry I did only local builds...

LocutusOfBorg replied:
are we good? do you want any more log?
```
/bin/bash ../../../libtool  --tag=CC   --mode=link gcc -g -O0   -pthread -Wall -Wextra -Wno-unused-parameter -Werror -Wstrict-aliasing=2 -g -O2 -ffile-prefix-map=/build/opensc-0.23.0=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -fdebug-prefix-map=/build/opensc-0.23.0=/usr/src/opensc-0.23.0-1ubuntu3 -Wno-error=deprecated-declarations -Wno-error=stringop-overflow  -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -Wl,-z,relro -o fuzz_piv_tool fuzz_piv_tool.o fuzzer_reader.o fuzzer_tool.o fuzzer.o ../../tools/util.o -lcrypto   ../../../src/libopensc/libopensc.la ../../../src/common/libscdl.la ../../../src/pkcs15init/libpkcs15init.la ../../../src/common/libcompat.la
/bin/bash ../../../libtool  --tag=CC   --mode=link gcc -g -O0   -pthread -Wall -Wextra -Wno-unused-parameter -Werror -Wstrict-aliasing=2 -g -O2 -ffile-prefix-map=/build/opensc-0.23.0=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -fdebug-prefix-map=/build/opensc-0.23.0=/usr/src/opensc-0.23.0-1ubuntu3 -Wno-error=deprecated-declarations -Wno-error=stringop-overflow -static -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -Wl,-z,relro -o fuzz_pkcs15init fuzz_pkcs15init.o fuzzer_reader.o fuzzer.o -lcrypto  -leac -lcrypto   ../../../src/libopensc/libopensc.la ../../../src/common/libscdl.la ../../../src/pkcs15init/libpkcs15init.la ../../../src/common/libcompat.la
libtool: link: gcc -g -O0 -Wall -Wextra -Wno-unused-parameter -Werror -Wstrict-aliasing=2 -g -O2 -ffile-prefix-map=/build/opensc-0.23.0=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -fdebug-prefix-map=/build/opensc-0.23.0=/usr/src/opensc-0.23.0-1ubuntu3 -Wno-error=deprecated-declarations -Wno-error=stringop-overflow -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -Wl,-z -Wl,relro -o .libs/fuzz_piv_tool fuzz_piv_tool.o fuzzer_reader.o fuzzer_tool.o fuzzer.o ../../tools/util.o  -lcrypto ../../../src/libopensc/.libs/libopensc.so ../../../src/common/.libs/libscdl.a -ldl ../../../src/pkcs15init/.libs/libpkcs15init.a ../../../src/common/.libs/libcompat.a -pthread
libtool: link: gcc -g -O0 -Wall -Wextra -Wno-unused-parameter -Werror -Wstrict-aliasing=2 -g -O2 -ffile-prefix-map=/build/opensc-0.23.0=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -fdebug-prefix-map=/build/opensc-0.23.0=/usr/src/opensc-0.23.0-1ubuntu3 -Wno-error=deprecated-declarations -Wno-error=stringop-overflow -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -Wl,-z -Wl,relro -o fuzz_pkcs15init fuzz_pkcs15init.o fuzzer_reader.o fuzzer.o  ../../../src/libopensc/.libs/libopensc.a -lz -lgio-2.0 -lgobject-2.0 -leac -lcrypto ../../../src/common/.libs/libscdl.a -ldl ../../../src/pkcs15init/.libs/libpkcs15init.a ../../../src/common/.libs/libcompat.a -pthread
../../pkcs15init/pkcs15-entersafe.c: In function 'entersafe_sanity_check':
../../pkcs15init/pkcs15-entersafe.c:497:33: error: 'profile_auth.attrs.pin.flags' may be used uninitialized [-Werror=maybe-uninitialized]
  497 |                                 && pin_attrs->flags != profile_auth.attrs.pin.flags)   {
      |                                 ^
../../pkcs15init/pkcs15-entersafe.c:475:36: note: 'profile_auth.attrs.pin.flags' was declared here
  475 |         struct sc_pkcs15_auth_info profile_auth;
      |                                    ^
../../pkcs15init/pkcs15-entersafe.c:496:20: error: 'profile_auth.attrs.pin.reference' may be used uninitialized [-Werror=maybe-uninitialized]
  496 |                 if (pin_attrs->reference == profile_auth.attrs.pin.reference
      |                    ^
../../pkcs15init/pkcs15-entersafe.c:475:36: note: 'profile_auth.attrs.pin.reference' was declared here
  475 |         struct sc_pkcs15_auth_info profile_auth;
      |                                    ^
../../pkcs15init/pkcs15-epass2003.c: In function 'epass2003_pkcs15_sanity_check':
../../pkcs15init/pkcs15-epass2003.c:756:21: error: 'profile_auth.attrs.pin.flags' may be used uninitialized [-Werror=maybe-uninitialized]
  756 |                     && pin_attrs->flags != profile_auth.attrs.pin.flags) {
      |                     ^
../../pkcs15init/pkcs15-epass2003.c:732:36: note: 'profile_auth.attrs.pin.flags' was declared here
  732 |         struct sc_pkcs15_auth_info profile_auth;
      |                                    ^
../../pkcs15init/pkcs15-epass2003.c:755:20: error: 'profile_auth.attrs.pin.reference' may be used uninitialized [-Werror=maybe-uninitialized]
  755 |                 if (pin_attrs->reference == profile_auth.attrs.pin.reference
      |                    ^
../../pkcs15init/pkcs15-epass2003.c:732:36: note: 'profile_auth.attrs.pin.reference' was declared here
  732 |         struct sc_pkcs15_auth_info profile_auth;
      |                                    ^
In function 'have_onepin',
    inlined from 'asepcos_create_pin' at ../../pkcs15init/pkcs15-asepcos.c:419:7:
../../pkcs15init/pkcs15-asepcos.c:325:37: error: 'sopin.attrs.pin.flags' may be used uninitialized [-Werror=maybe-uninitialized]
  325 |         if (!(sopin.attrs.pin.flags & SC_PKCS15_PIN_FLAG_SO_PIN))
      |                                     ^
../../pkcs15init/pkcs15-asepcos.c: In function 'asepcos_create_pin':
../../pkcs15init/pkcs15-asepcos.c:321:31: note: 'sopin.attrs.pin.flags' was declared here
  321 |         sc_pkcs15_auth_info_t sopin;
      |                               ^
../../pkcs15init/pkcs15-asepcos.c: In function 'have_onepin':
../../pkcs15init/pkcs15-asepcos.c:325:37: error: 'sopin.attrs.pin.flags' may be used uninitialized [-Werror=maybe-uninitialized]
  325 |         if (!(sopin.attrs.pin.flags & SC_PKCS15_PIN_FLAG_SO_PIN))
      |                                     ^
../../pkcs15init/pkcs15-asepcos.c:321:31: note: 'sopin.attrs.pin.flags' was declared here
  321 |         sc_pkcs15_auth_info_t sopin;
      |                               ^
lto1: all warnings being treated as errors
make[6]: *** [/tmp/ccRMbfeN.mk:71: /tmp/ccxVJcQw.ltrans23.ltrans.o] Error 1
make[6]: *** Waiting for unfinished jobs....

```
this is the error with that variable non memset to 0

Jakuje replied:
Thank you for the logs! I think the memset here is more like patch hiding the other problem that the callers of `sc_profile_get_pin_info()` do not initialize the `info` argument and do not check the return values of this function, causing uninitialized reads.

My preference would be either or both
 * Initialize the `info` argument in all calls to this function
 * Check return values of this function before using its results.

frankmorgner commented:
What version of OpenSC are you operating on?

The changes for initializing `apdu_t` should not be needed, because `sc_format_apdu` already does `memset` this memory and it is always called from the lines you have modified. The only possibility how this could yield a warning would be if one of the card driver functions was called `card` set to `NULL` (in this case `sc_format_apdu` exits early). However, the card driver functions are always used with a valid `card` object. If you're in master, then this seems to be a false positive...

LocutusOfBorg commented:
> What version of OpenSC are you operating on?
0.23.0

Jakuje commented on src/pkcs15init/profile.c:
```diff
@@ -499,6 +499,7 @@ sc_profile_get_pin_info(struct sc_profile *profile,
 		int id, struct sc_pkcs15_auth_info *info)
 {
 	struct pin_info	*pi;
+	memset(info, 0, sizeof(*info));
```
Posted to the previous version of the comments inline, as it still applies:

I think the memset here is more like patch hiding the other problem that the callers of `sc_profile_get_pin_info()` do not initialize the info argument and do not check the return values of this function, causing uninitialized reads.

My preference would be either or both of the following instead of blindly calling memset on a argument of the function:

* Initialize the info argument in all calls to this function
* Check return values of this function before using its results.

LocutusOfBorg replied:
ok, removed. I added some more initialization to fix other failures due to the missing memset, as you suggested.

LocutusOfBorg commented:
@frankmorgner yes, it is a false positive, but gcc is not that smart, and initializing the memory to 0 is a sane thing to do...
Updated the pr, removed some check and initialized the variables where needed.

Jakuje commented on src/pkcs15init/pkcs15-lib.c:
```diff
@@ -1078,6 +1081,9 @@ sc_pkcs15init_store_pin(struct sc_pkcs15_card *p15card, struct sc_profile *profi
 	auth_info = (struct sc_pkcs15_auth_info *) pin_obj->data;
 
 	sc_profile_get_pin_info(profile, SC_PKCS15INIT_USER_PIN, auth_info);
+	if (auth_info == NULL )
```
nit
```suggestion
	if (auth_info == NULL)
```

Jakuje commented on src/pkcs15init/pkcs15-lib.c:
```diff
@@ -1014,6 +1014,9 @@ sc_pkcs15init_store_puk(struct sc_pkcs15_card *p15card,
 	auth_info = (struct sc_pkcs15_auth_info *) pin_obj->data;
 
 	sc_profile_get_pin_info(profile, SC_PKCS15INIT_USER_PUK, auth_info);
+	if (auth_info == NULL )
```
nit
```suggestion
	if (auth_info == NULL)
```

Jakuje reviewed (approved):
aside of minor formatting nits, looks good

LocutusOfBorg commented:
thanks!

LocutusOfBorg commented:
I rebased hoping to not trigger any ci failure again

frankmorgner reviewed (approved):
thank you

LocutusOfBorg commented:
now everything is green!

Jakuje commented:
Thank you! Merging.
